File: apps/cargo-tracker/src/main/java/com/example/cargotracker/booking/domain/model/aggregates/Cargo.java
package com.example.cargotracker.booking.domain.model.aggregates;

import com.example.cargotracker.booking.domain.model.valueobjects.BookingId;
import com.example.cargotracker.booking.domain.model.valueobjects.Description;
import com.example.cargotracker.booking.domain.model.valueobjects.Dimensions;
import com.example.cargotracker.booking.domain.model.valueobjects.HazardousDeclaration;
import com.example.cargotracker.booking.domain.model.valueobjects.Quantity;
import com.example.cargotracker.booking.domain.model.valueobjects.RouteSpecification;
import com.example.cargotracker.booking.domain.model.valueobjects.TemperatureRequirement;
import com.example.cargotracker.shared.domain.model.ShipperId;

import java.math.BigDecimal;
import java.util.EnumSet;
import java.util.Objects;

public class Cargo {

    private static final EnumSet<BookingStatus> CANCELLABLE_STATUSES =
            EnumSet.of(BookingStatus.PRELIMINARY, BookingStatus.ROUTE_PROPOSED, BookingStatus.CONFIRMED,
                    BookingStatus.TRACKING_ISSUED, BookingStatus.IN_TRANSIT);

    private final BookingId bookingId;
    private final ShipperId shipperId;
    private final CargoType cargoType;
    private final BigDecimal weight;
    private final Dimensions dimensions;
    private final Quantity quantity;
    private final Description description;
    private final RouteSpecification routeSpecification;
    private final BookingStatus status;
    private final HazardousDeclaration hazardousDeclaration;
    private final TemperatureRequirement temperatureRequirement;

    public record Details(
            Dimensions dimensions,
            Quantity quantity,
            Description description
    ) {
    }

    public record Handling(
            HazardousDeclaration hazardousDeclaration,
            TemperatureRequirement temperatureRequirement
    ) {
    }

    public record State(
            RouteSpecification routeSpecification,
            BookingStatus status,
            Details details,
            Handling handling
    ) {
    }

    public static Details details(Dimensions dimensions, Quantity quantity, Description description) {
        return new Details(dimensions, quantity, description);
    }

    public static Handling handling(
            HazardousDeclaration hazardousDeclaration,
            TemperatureRequirement temperatureRequirement
    ) {
        return new Handling(hazardousDeclaration, temperatureRequirement);
    }

    public static State state(
            RouteSpecification routeSpecification,
            BookingStatus status,
            Details details,
            Handling handling
    ) {
        return new State(routeSpecification, status, details, handling);
    }

    public Cargo(
            BookingId bookingId,
            ShipperId shipperId,
            CargoType cargoType,
            BigDecimal weight,
            RouteSpecification routeSpecification
    ) {
        this(bookingId, shipperId, cargoType, weight, state(routeSpecification, BookingStatus.PRELIMINARY, null, null));
    }

    public Cargo(
            BookingId bookingId,
            ShipperId shipperId,
            CargoType cargoType,
            BigDecimal weight,
            RouteSpecification routeSpecification,
            BookingStatus status
    ) {
        this(bookingId, shipperId, cargoType, weight, state(routeSpecification, status, null, null));
    }

    public Cargo(
            BookingId bookingId,
            ShipperId shipperId,
            CargoType cargoType,
            BigDecimal weight,
            State state
    ) {
        this.bookingId = Objects.requireNonNull(bookingId, "bookingId must not be null");
        this.shipperId = Objects.requireNonNull(shipperId, "shipperId must not be null");
        this.cargoType = Objects.requireNonNull(cargoType, "cargoType must not be null");
        this.weight = Objects.requireNonNull(weight, "weight must not be null");
        State cargoState = Objects.requireNonNull(state, "state must not be null");
        this.dimensions = cargoState.details() != null ? cargoState.details().dimensions() : null;
        this.quantity = cargoState.details() != null ? cargoState.details().quantity() : null;
        this.description = cargoState.details() != null ? cargoState.details().description() : null;
        this.routeSpecification = Objects.requireNonNull(cargoState.routeSpecification(), "routeSpecification must not be null");
        this.status = Objects.requireNonNull(cargoState.status(), "status must not be null");
        this.hazardousDeclaration = cargoState.handling() != null ? cargoState.handling().hazardousDeclaration() : null;
        this.temperatureRequirement = cargoState.handling() != null ? cargoState.handling().temperatureRequirement() : null;

        if (this.weight.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("weight must be greater than zero");
        }
        if (this.cargoType == CargoType.HAZARDOUS && this.hazardousDeclaration == null) {
            throw new IllegalArgumentException("hazardousDeclaration is required for HAZARDOUS cargo");
        }
        if (this.cargoType == CargoType.REFRIGERATED && this.temperatureRequirement == null) {
            throw new IllegalArgumentException("temperatureRequirement is required for REFRIGERATED cargo");
        }
    }

    public BookingId getBookingId() {
        return bookingId;
    }

    public ShipperId getShipperId() {
        return shipperId;
    }

    public CargoType getCargoType() {
        return cargoType;
    }

    public BigDecimal getWeight() {
        return weight;
    }

    public Dimensions getDimensions() {
        return dimensions;
    }

    public Quantity getQuantity() {
        return quantity;
    }

    public Description getDescription() {
        return description;
    }

    public RouteSpecification getRouteSpecification() {
        return routeSpecification;
    }

    public BookingStatus getStatus() {
        return status;
    }

    public HazardousDeclaration getHazardousDeclaration() {
        return hazardousDeclaration;
    }

    public TemperatureRequirement getTemperatureRequirement() {
        return temperatureRequirement;
    }

    public Cargo confirm() {
        if (status != BookingStatus.PRELIMINARY) {
            throw new IllegalStateException("予約確定は仮受付状態からのみ可能です。現在の状態: " + status.getDisplayName());
        }
        return new Cargo(bookingId, shipperId, cargoType, weight, currentStateWith(BookingStatus.CONFIRMED));
    }

    public Cargo cancel() {
        if (!CANCELLABLE_STATUSES.contains(status)) {
            throw new IllegalStateException("現在の状態ではキャンセルできません。現在の状態: " + status.getDisplayName());
        }
        return new Cargo(bookingId, shipperId, cargoType, weight, currentStateWith(BookingStatus.CANCELLED));
    }

    public Cargo assignToRouting() {
        if (status != BookingStatus.PRELIMINARY) {
            throw new IllegalStateException("経路設計依頼は仮受付状態からのみ可能です。現在の状態: " + status.getDisplayName());
        }
        return new Cargo(bookingId, shipperId, cargoType, weight, currentStateWith(BookingStatus.ROUTE_PROPOSED));
    }

    private State currentStateWith(BookingStatus nextStatus) {
        return state(
                routeSpecification,
                nextStatus,
                details(dimensions, quantity, description),
                handling(hazardousDeclaration, temperatureRequirement)
        );
    }
}

File: apps/cargo-tracker/src/CargoTracker.Web/Booking/Domain/Model/Cargo.cs
using CargoTracker.Booking.Domain.Events;
using CargoTracker.Shared.Domain.Model;

namespace CargoTracker.Booking.Domain.Model;

/// <summary>貨物予約集約ルート（US04）。</summary>
public sealed class Cargo : AggregateRoot
{
    public BookingId BookingId { get; }
    public ShipperId ShipperId { get; }
    public RouteSpecification RouteSpecification { get; }
    public CargoType CargoType { get; }
    public decimal Weight { get; }
    public Dimensions? Dimensions { get; }
    public Quantity? Quantity { get; }
    public Description? Description { get; }
    public HazardousDeclaration? HazardousDeclaration { get; }
    public TemperatureRequirement? TemperatureRequirement { get; }
    public BookingStatus BookingStatus { get; private set; }
    public CargoItinerary? CargoItinerary { get; private set; }
    public long Version { get; private set; }

    private Cargo(
        BookingId bookingId, ShipperId shipperId, RouteSpecification routeSpecification, CargoType cargoType,
        decimal weight, Dimensions? dimensions, Quantity? quantity, Description? description,
        HazardousDeclaration? hazardousDeclaration, TemperatureRequirement? temperatureRequirement,
        BookingStatus bookingStatus, long version, CargoItinerary? cargoItinerary = null)
    {
        BookingId = bookingId;
        ShipperId = shipperId;
        RouteSpecification = routeSpecification;
        CargoType = cargoType;
        Weight = weight;
        Dimensions = dimensions;
        Quantity = quantity;
        Description = description;
        HazardousDeclaration = hazardousDeclaration;
        TemperatureRequirement = temperatureRequirement;
        BookingStatus = bookingStatus;
        CargoItinerary = cargoItinerary;
        Version = version;
    }

    public static Cargo Create(
        ShipperId shipperId, RouteSpecification routeSpecification, CargoType cargoType, decimal weight,
        Dimensions? dimensions = null, Quantity? quantity = null, Description? description = null,
        HazardousDeclaration? hazardousDeclaration = null, TemperatureRequirement? temperatureRequirement = null,
        DateOnly? today = null)
    {
        if (weight <= 0)
        {
            throw new ArgumentException("重量は正の値でなければなりません。", nameof(weight));
        }
        // 到着期限は当日以降でなければならない（IT2 レビュー H3。Estimate.Create と同じ不変条件）。
        var currentDate = today ?? DateOnly.FromDateTime(DateTime.UtcNow);
        if (routeSpecification.ArrivalDeadline < currentDate)
        {
            throw new ArgumentException("到着期限は当日以降でなければなりません。", nameof(routeSpecification));
        }
        ValidateSpecialRequirements(cargoType, hazardousDeclaration, temperatureRequirement);

        var cargo = new Cargo(
            BookingId.Generate(), shipperId, routeSpecification, cargoType, weight,
            dimensions, quantity, description, hazardousDeclaration, temperatureRequirement, BookingStatus.Preliminary, 0);
        cargo.AddDomainEvent(new CargoBookedEvent(cargo.BookingId));
        return cargo;
    }

    public void AssignToRouting()
    {
        if (BookingStatus != BookingStatus.Preliminary)
        {
            throw new InvalidOperationException("仮受付の予約のみ経路設計へ割り当てられます。");
        }
        BookingStatus = BookingStatus.RouteProposed;
        Version++;
        AddDomainEvent(new AssignedToRoutingEvent(BookingId));
    }

    /// <summary>確定経路（旅程）を予約に紐付ける（US11）。状態は RouteProposed のまま維持する。</summary>
    public void AssignItinerary(CargoItinerary itinerary)
    {
        ArgumentNullException.ThrowIfNull(itinerary);
        if (BookingStatus != BookingStatus.RouteProposed)
        {
            throw new InvalidOperationException("経路提案中の予約のみ経路を紐付けられます。");
        }
        CargoItinerary = itinerary;
        Version++;
        AddDomainEvent(new CargoRoutedEvent(BookingId));
    }

    /// <summary>予約を確定する（US13）。経路提案中かつ旅程割当済みでなければならない。</summary>
    public void Confirm()
    {
        if (BookingStatus != BookingStatus.RouteProposed)
        {
            throw new InvalidOperationException("経路提案中の予約のみ確定できます。");
        }
        if (CargoItinerary is null)
        {
            throw new InvalidOperationException("経路が紐付けられていない予約は確定できません。");
        }
        BookingStatus = BookingStatus.Confirmed;
        Version++;
        AddDomainEvent(new BookingConfirmedEvent(BookingId));
    }

    /// <summary>荷主のルート変更希望で経路再設計に差し戻す（US13）。RouteProposed → Preliminary。</summary>
    public void ReturnToRouting()
    {
        if (BookingStatus != BookingStatus.RouteProposed)
        {
            throw new InvalidOperationException("経路提案中の予約のみ経路再設計に差し戻せます。");
        }
        BookingStatus = BookingStatus.Preliminary;
        CargoItinerary = null;
        Version++;
    }

    /// <summary>予約をキャンセルする（US13）。確定後・終端状態からはキャンセルできない。</summary>
    public void Cancel()
    {
        if (BookingStatus is not (BookingStatus.Preliminary or BookingStatus.RouteProposed))
        {
            throw new InvalidOperationException("仮受付または経路提案中の予約のみキャンセルできます。");
        }
        BookingStatus = BookingStatus.Cancelled;
        Version++;
    }

    /// <summary>永続化データから集約を再構築する（イベントは発生させない）。</summary>
    public static Cargo Reconstruct(
        BookingId bookingId, ShipperId shipperId, RouteSpecification routeSpecification, CargoType cargoType,
        decimal weight, Dimensions? dimensions, Quantity? quantity, Description? description,
        BookingStatus bookingStatus, long version,
        HazardousDeclaration? hazardousDeclaration = null, TemperatureRequirement? temperatureRequirement = null,
        CargoItinerary? cargoItinerary = null)
        => new(bookingId, shipperId, routeSpecification, cargoType, weight, dimensions, quantity, description,
            hazardousDeclaration, temperatureRequirement,
            bookingStatus, version, cargoItinerary);

    private static void ValidateSpecialRequirements(
        CargoType cargoType, HazardousDeclaration? hazardousDeclaration, TemperatureRequirement? temperatureRequirement)
    {
        switch (cargoType)
        {
            case CargoType.Hazardous:
                if (hazardousDeclaration is null)
                {
                    throw new ArgumentException("危険物貨物には危険物申告が必要です。", nameof(hazardousDeclaration));
                }
                if (temperatureRequirement is not null)
                {
                    throw new ArgumentException("危険物貨物に温度管理条件は指定できません。", nameof(temperatureRequirement));
                }
                break;
            case CargoType.Refrigerated:
                if (temperatureRequirement is null)
                {
                    throw new ArgumentException("冷凍・冷蔵貨物には温度管理条件が必要です。", nameof(temperatureRequirement));
                }
                if (hazardousDeclaration is not null)
                {
                    throw new ArgumentException("冷凍・冷蔵貨物に危険物申告は指定できません。", nameof(hazardousDeclaration));
                }
                break;
            case CargoType.General:
                if (hazardousDeclaration is not null || temperatureRequirement is not null)
                {
                    throw new ArgumentException("一般貨物に特別情報は指定できません。");
                }
                break;
            default:
                throw new ArgumentOutOfRangeException(nameof(cargoType), cargoType, "未対応の貨物種別です。");
        }
    }
}
